Support additional unpacking generalisations (PEP 448)

I just would like to add that although the new parser (mentioned in #1353) parses below:

```
my_dict = {'b': 2, 'c': 3}
my_new_dict = {'a': 1, **my_dict}
```

mypy ends up in a bad place:
```
$ python3.5 -m mypy --fast-parser mypy_pep0448.py
Traceback (most recent call last):
  File "/home/mb/.local/lib/python3.5/runpy.py", line 170, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/mb/.local/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/__main__.py", line 5, in <module>
    main(None)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/main.py", line 54, in main
    res = type_check_only(sources, bin_dir, options)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/main.py", line 102, in type_check_only
    python_path=options.python_path)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/build.py", line 211, in build
    dispatch(sources, manager)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/build.py", line 1334, in dispatch
    graph = load_graph(sources, manager)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/build.py", line 1349, in load_graph
    st = State(id=bs.module, path=bs.path, source=bs.text, manager=manager)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/build.py", line 1121, in __init__
    self.parse_file()
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/build.py", line 1232, in parse_file
    self.tree = manager.parse_file(self.id, self.xpath, source)
  File "/home/mb/.local/lib/python3.5/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/build.py", line 1186, in wrap_context
    yield
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/build.py", line 1232, in parse_file
    self.tree = manager.parse_file(self.id, self.xpath, source)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/build.py", line 452, in parse_file
    fast_parser=FAST_PARSER in self.flags)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/parse.py", line 88, in parse
    custom_typing_module=custom_typing_module)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/fastparse.py", line 66, in parse
    ).visit(ast)
  File "/home/mb/.local/lib/python3.5/site-packages/typed_ast/ast35.py", line 252, in visit
    return visitor(node)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/fastparse.py", line 228, in visit_Module
    body = self.fix_function_overloads(self.visit_list(mod.body))
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/fastparse.py", line 129, in visit_list
    return [self.visit(e) for e in l]
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/fastparse.py", line 129, in <listcomp>
    return [self.visit(e) for e in l]
  File "/home/mb/.local/lib/python3.5/site-packages/typed_ast/ast35.py", line 252, in visit
    return visitor(node)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/fastparse.py", line 97, in wrapper
    node = f(self, ast)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/fastparse.py", line 387, in visit_Assign
    self.visit(n.value),
  File "/home/mb/.local/lib/python3.5/site-packages/typed_ast/ast35.py", line 252, in visit
    return visitor(node)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/fastparse.py", line 97, in wrapper
    node = f(self, ast)
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/fastparse.py", line 574, in visit_Dict
    return DictExpr(list(zip(self.visit_list(n.keys), self.visit_list(n.values))))
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/nodes.py", line 1445, in __init__
    for x in items):
  File "/home/mb/.local/lib/python3.5/site-packages/mypy/nodes.py", line 1445, in <genexpr>
    for x in items):
AttributeError: 'NoneType' object has no attribute 'literal'
```

for reference, using current parser:

```
$ python3.5 -m mypy mypy_pep0448.py
mypy_pep0448.py:2: error: Parse error before **
```

also for reference: Ubuntu 14.04 x64, Python 3.5.1, mypy 0.4.2, typed-ast 0.5.2

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -1360,29 +1360,58 @@ class ExpressionChecker:
         return TupleType(items, self.chk.named_generic_type('builtins.tuple', [fallback_item]))
 
     def visit_dict_expr(self, e: DictExpr) -> Type:
-        # Translate into type checking a generic function call.
+        """Type check a dict expression.
+
+        Translate it into a call to dict(), with provisions for **expr.
+        """
+        # Collect function arguments, watching out for **expr.
+        args = []  # type: List[Node]  # Regular "key: value"
+        stargs = []  # type: List[Node]  # For "**expr"
+        for key, value in e.items:
+            if key is None:
+                stargs.append(value)
+            else:
+                args.append(TupleExpr([key, value]))
+        # Define type variables (used in constructors below).
         ktdef = TypeVarDef('KT', -1, [], self.chk.object_type())
         vtdef = TypeVarDef('VT', -2, [], self.chk.object_type())
         kt = TypeVarType(ktdef)
         vt = TypeVarType(vtdef)
-        # The callable type represents a function like this:
-        #
-        #   def <unnamed>(*v: Tuple[kt, vt]) -> Dict[kt, vt]: ...
-        constructor = CallableType(
-            [TupleType([kt, vt], self.named_type('builtins.tuple'))],
-            [nodes.ARG_STAR],
-            [None],
-            self.chk.named_generic_type('builtins.dict', [kt, vt]),
-            self.named_type('builtins.function'),
-            name='<list>',
-            variables=[ktdef, vtdef])
-        # Synthesize function arguments.
-        args = []  # type: List[Node]
-        for key, value in e.items:
-            args.append(TupleExpr([key, value]))
-        return self.check_call(constructor,
-                               args,
-                               [nodes.ARG_POS] * len(args), e)[0]
+        # Call dict(*args), unless it's empty and stargs is not.
+        if args or not stargs:
+            # The callable type represents a function like this:
+            #
+            #   def <unnamed>(*v: Tuple[kt, vt]) -> Dict[kt, vt]: ...
+            constructor = CallableType(
+                [TupleType([kt, vt], self.named_type('builtins.tuple'))],
+                [nodes.ARG_STAR],
+                [None],
+                self.chk.named_generic_type('builtins.dict', [kt, vt]),
+                self.named_type('builtins.function'),
+                name='<list>',
+                variables=[ktdef, vtdef])
+            rv = self.check_call(constructor, args, [nodes.ARG_POS] * len(args), e)[0]
+        else:
+            # dict(...) will be called below.
+            rv = None
+        # Call rv.update(arg) for each arg in **stargs,
+        # except if rv isn't set yet, then set rv = dict(arg).
+        if stargs:
+            for arg in stargs:
+                if rv is None:
+                    constructor = CallableType(
+                        [self.chk.named_generic_type('typing.Mapping', [kt, vt])],
+                        [nodes.ARG_POS],
+                        [None],
+                        self.chk.named_generic_type('builtins.dict', [kt, vt]),
+                        self.named_type('builtins.function'),
+                        name='<list>',
+                        variables=[ktdef, vtdef])
+                    rv = self.check_call(constructor, [arg], [nodes.ARG_POS], arg)[0]
+                else:
+                    method = self.analyze_external_member_access('update', rv, arg)
+                    self.check_call(method, [arg], [nodes.ARG_POS], arg)
+        return rv
 
     def visit_func_expr(self, e: FuncExpr) -> Type:
         """Type check lambda expression."""

--- a/mypy/nodes.py
+++ b/mypy/nodes.py
@@ -1457,7 +1457,9 @@ class DictExpr(Expression):
 
     def __init__(self, items: List[Tuple[Expression, Expression]]) -> None:
         self.items = items
-        if all(x[0].literal == LITERAL_YES and x[1].literal == LITERAL_YES
+        # key is None for **item, e.g. {'a': 1, **x} has
+        # keys ['a', None] and values [1, x].
+        if all(x[0] and x[0].literal == LITERAL_YES and x[1].literal == LITERAL_YES
                for x in items):
             self.literal = LITERAL_YES
             self.literal_hash = ('Dict',) + tuple(

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -1937,7 +1937,8 @@ class SemanticAnalyzer(NodeVisitor):
 
     def visit_dict_expr(self, expr: DictExpr) -> None:
         for key, value in expr.items:
-            key.accept(self)
+            if key is not None:
+                key.accept(self)
             value.accept(self)
 
     def visit_star_expr(self, expr: StarExpr) -> None:

--- a/mypy/traverser.py
+++ b/mypy/traverser.py
@@ -180,7 +180,8 @@ class TraverserVisitor(NodeVisitor[None]):
 
     def visit_dict_expr(self, o: DictExpr) -> None:
         for k, v in o.items:
-            k.accept(self)
+            if k is not None:
+                k.accept(self)
             v.accept(self)
 
     def visit_set_expr(self, o: SetExpr) -> None:

--- a/test-data/unit/check-expressions.test
+++ b/test-data/unit/check-expressions.test
@@ -1503,3 +1503,19 @@ None == None
 [case testLtNone]
 None < None  # E: Unsupported left operand type for < (None)
 [builtins fixtures/ops.py]
+
+[case testDictWithStarExpr]
+# options: fast_parser
+b = {'z': 26, *a}  # E: invalid syntax
+[builtins fixtures/dict.py]
+
+[case testDictWithStarStarExpr]
+# options: fast_parser
+from typing import Dict
+a = {'a': 1}
+b = {'z': 26, **a}
+c = {**b}
+d = {**a, **b, 'c': 3}
+e = {1: 'a', **a}  # E: Argument 1 to "update" of "dict" has incompatible type Dict[str, int]; expected Mapping[int, str]
+f = {**b}  # type: Dict[int, int]  # E: List item 0 has incompatible type Dict[str, int]
+[builtins fixtures/dict.py]

--- a/test-data/unit/fixtures/dict.py
+++ b/test-data/unit/fixtures/dict.py
@@ -1,6 +1,6 @@
 # Builtins stub used in dictionary-related test cases.
 
-from typing import TypeVar, Generic, Iterable, Iterator, Tuple, overload
+from typing import TypeVar, Generic, Iterable, Iterator, Mapping, Tuple, overload
 
 T = TypeVar('T')
 KT = TypeVar('KT')
@@ -11,14 +11,14 @@ class object:
 
 class type: pass
 
-class dict(Iterable[KT], Generic[KT, VT]):
+class dict(Iterable[KT], Mapping[KT, VT], Generic[KT, VT]):
     @overload
     def __init__(self, **kwargs: VT) -> None: pass
     @overload
     def __init__(self, arg: Iterable[Tuple[KT, VT]], **kwargs: VT) -> None: pass
     def __setitem__(self, k: KT, v: VT) -> None: pass
     def __iter__(self) -> Iterator[KT]: pass
-    def update(self, a: 'dict[KT, VT]') -> None: pass
+    def update(self, a: Mapping[KT, VT]) -> None: pass
 
 class int: pass # for convenience
 

--- a/test-data/unit/lib-stub/typing.py
+++ b/test-data/unit/lib-stub/typing.py
@@ -74,6 +74,8 @@ class Sequence(Iterable[T], Generic[T]):
     @abstractmethod
     def __getitem__(self, n: Any) -> T: pass
 
+class Mapping(Generic[T, U]): pass
+
 def NewType(name: str, tp: Type[T]) -> Callable[[T], T]:
     def new_type(x):
         return x

